### ISSUE TITLE
Changing the title of a page or post results in that URL appearing twice in the "Most Popular Content" table

### ISSUE THREAD

**Issue by jamesozzie** [google/site-kit-wp#3070]
## Bug Description

When making changes such as the following this results in the "Most popular content" table listing duplicate content:
- Post or page title (without any slug or permalink changes)
- Site Title changes (Settings > General)
- Slug changes

The changes above reset the statistics for that content, with that same URL appearing twice in the "Most Popular Content" table - Once with total views based on the previous title and again with the new title. 

Any changes to a sites title will result in the homepage being listed twice

![image](https://user-images.githubusercontent.com/41326532/113275967-3df4a800-92d7-11eb-952b-c4bac42f78fb.png)

## Steps to reproduce

1. Change the title or a page or post
2. Don't make any slug changes
3. Revisit the most popular content table after traffic to those pages
4. One occurrence will appear prior to the URL change, another occurrence for after the title change

## Additional Context

 - Duplicate references do not occur on Analytics directly
 - Separately any slug changes will also result in duplicate occurrences in the same table (Tested when the post is part of the permalink structure)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* All _refactored_ Analytics widgets which display most popular/most earning titles and URLs (e.g. in WP dashboard widget, Popular Pages widget and Top Earning Pages widgets on SK dashboard, the same widgets on Analytics/AdSense module page) should be modified so that they don't consider a URL multiple records just because the URL's title may have changed over time. This is currently resulting in:
    * slightly incorrect numbers compared to the Analytics frontend (e.g. for one title the URL has 1000 views, while for the other title it has 4, which means in Site Kit probably only the first record makes it to the list, showing 1000, while the Analytics frontend shows 1004)
    * duplicate results, particularly on sites with low traffic (e.g. for one title the URL has 5 views, while for the other title it has 4, which means in Site Kit probably both records show in the list, but it's the same URL, while in Analytics frontend it's just one record showing 9 views)
* The main technical problem here is that what we actually want is some sort of Analytics metric like `ga:latestPageTitle`, which however doesn't exist. There is no way to get the titles from Analytics without also separating it out as a dimension - but we need it as a metric. Therefore we need a workaround, as follows:
    * Change the primary Analytics API request that is used to only query by the `ga:pagePath` dimension, instead of also including the `ga:pageTitle` dimension. This ensures that results will be grouped purely by the URL/path.
    * Then, internally an additional Analytics request with all the same arguments, except that _that_ request should still use both `ga:pagePath` and `ga:pageTitle`, _and_ the number of items requested should be 5 times the number of what is _actually_ being requested (e.g. if the table shows 10 results, request the top 50 in that second request).
    * The response from that second request should then be used to match URLs from the first response with titles present in the second response for each URL.
        * Always use the first title found for the URL (i.e. if there are multiple titles, ignore the following ones).
        * If despite the high number of entries in the second response no title is found for a URL (unlikely but possible), fall back to using "(unknown)" as title.

## Implementation Brief

* Update the `isValidDimensionFilters` function to allow dimension value to be an array of string or number values.
* Update the dimension filter section in the `GET:report` case of the `Analytics::create_data_request` method:
  * Set filter operator to `IN_LIST` when the `$dimension_value` variable is an array, otherwise keep using `EXACT` operator.
  * Do not wrap the `$dimension_value` variable into an array when setting expressions for the filter if the `$dimension_value` variable is already an array.
* Add a new selector `getPageTitles( pagePaths, startDate, endDate )` to the reports datastore of the Analytics module that returns a `pagePath => pageTitle` map.
  * Internally this selector should call the `getReport` selector with the following arguments:
    * Start and end dates should use corresponding `startDate` and `endDate` values.
    * The `metrics` array should contain only the `ga:pageviews` value.
    * The `dimensions` list should include `ga:pagePaths` and `ga:pageTitle` values.
    * The `dimensionFilter` map should contain `ga:pagePath => {array of incoming page paths}`
    * The `limit` value should be five times bigger than the total count of page paths.
  * Return `undefined` if the `getReport` selector returns `undefined`
  * If the `getReport` selector returns valid report data, then loop through each row in the report and compose an object that consists of page paths and page titles. Use the first returned titles and don't override it if we already have a title for the page path.
  * Return composed object.
* Update the `DashboardPopularPagesWidget` component:
  * Remove the `ga:pageTitle` dimension from the report arguments.
  * Call the `getPageTitles` selector once the `getReport` selector is resolved and returns valid data. For the `getPageTitles` selector path page paths returned by the `getReport` selector, and startDate/endDate used in report arguments.
  * When the `getPageTitles` request is resolved and it returns valid data, update the `ReportTable` columns to use the newly aggregated page titles map to display titles.
* Use the similar approach for `ModulePopularPagesWidget`, `ModuleTopEarningPagesWidget`, and `WPDashboardPopularPages` components to get aggregated data.

### Test Coverage

* Add js tests for the new selector.

### Visual Regression Changes

* N/A

## QA Brief

0. Enable analytics,
1. Ensure you have some pages with same url and different titles, you can do this by publishing a page, triggering some hits in analytics, then changing its title but not url and triggering some more hits in analytics.
2. Open each of the components below.
3. Ensure each url is entered only once, unlike in the screenshot in bug description.

**DashboardPopularPagesWidget**
/wp-admin/admin.php?page=googlesitekit-dashboard

**ModuleTopEarningPagesWidget** (needs adsense enabled)
/wp-admin/admin.php?page=googlesitekit-module-adsense

**ModulePopularPagesWidget**
/wp-admin/admin.php?page=googlesitekit-module-analytics

**WPDashboardPopularPages**
/wp-admin/

Example Before:
![3070-before](https://user-images.githubusercontent.com/373931/131470484-1a666132-28de-4f18-a810-520f8a98b650.png)

Example After:
![3070-after](https://user-images.githubusercontent.com/373931/131470520-2595c15d-8359-4546-8eee-2dd68cfa5f65.png)

## Changelog entry

* Fix Analytics queries for URL lists to ignore titles for the metrics, while still showing the title for each URL.

**Comment by jamesozzie** [comment]
Additional insights after title change on a separate domain.

![image](https://user-images.githubusercontent.com/41326532/113410179-b24c4b80-93aa-11eb-8dfb-74fd1e21a4bd.png)

**Comment by jamesozzie** [comment]
@felixarntz @adamsilverstein Before closing this issue as the behaviour is expected, should a title change alone result in duplicate listings in the "Most Popular Content" table? Given the listings are pulled from Analytics and the final URL, total views and the "Detailed Page View" point to the same destination (as per [this demo](https://recordit.co/dn1FTibcNO))

**Comment by adamsilverstein** [comment]
I would expect them to since the list shows posts by title, not URL/slug. 

More importantly though, how do these URLs show up the the Search Console product UI? are they duplicated there as well? We should probably match that behavior.

**Comment by jamesozzie** [comment]
In Search Console (and Analytics) they are per URL. There is no "Title" heading in either product. 

![image](https://user-images.githubusercontent.com/41326532/116686965-68a45000-a9ac-11eb-8731-25b033a8b2d3.png)

I'll add for discussion next week.

**Comment by adamsilverstein** [comment]
Ah right, thanks for checking @jamesozzie. One potential way to resolve this that would better align with what user's see in Search console would be to group the most popular content by slug/url, then just show the most recent post title. 

From a user perspective, I feel like showing the page titles is more valuable/easier to understand than just showing the slug/url. At the same time, it is important that users see the same overall data when viewing data in Site Kit and the Search Console UI. As is, these are diverging because each changed title becomes a unique entry in Site Kit.

**Comment by jamesozzie** [comment]
I agree, displaying the page title along with the slug is ideal however if a user change's a site title could the listing in the most popular content table only show a URL one (with the latest title), in particular as both click through links land on the same "Detailed Page Stats" per the same URL

![image](https://user-images.githubusercontent.com/41326532/116712388-b7151700-a9cb-11eb-81bd-22708f4a4f38.png)

**Comment by felixarntz** [comment]
Yeah from an API perspective we should probably just query by URL. Let's see if it is possible though to also get back the (latest) title from Analytics, without that affecting the numbers/filtering though.

**Comment by jamesozzie** [comment]
[Here's another example](https://recordit.co/rPjGbOozSV) (based on another users seutp) showing the homepage title changed twice, therefore the homepage is listed 3 times. Different total views in the table, while all 3 "Detailed page views" show the total page views matching based on one of the listings.

**Comment by felixarntz** [comment]
Querying by only the `ga:pagePath` dimension would fix the duplicate results (and also the potentially incorrect numbers due to that!), however it doesn't look like it is possible to get the `ga:pageTitle` in that case. We could try to make a WordPress query to get the latest title from within WordPress, but that would be rather error-prone and also slow down performance.

@marrrmarrr We should consider only displaying the URLs here which would allow us to resolve this problem without introducing another error surface. While this does mean surfacing less information at a glance (since the titles would no longer be shown), it would be in parity with the Analytics frontend which doesn't show titles either by default (and therefore doesn't have to deal with this problem).

**Comment by felixarntz** [comment]
Per discussion with @marrrmarrr, we're going to solve this as follows:
* Make two API requests to Analytics, one for the top 10 URLs (only `ga:pagePath`) and another one for the top 50 URL + title combinations (`ga:pagePath` and `ga:pageTitle`, like now).
* The `ga:pagePath`-only request is the primary one which we'll use to populate the table. However, we'll amend the API response, combining it with the secondary response, taking from that one the first title that we find for the respective URL.
* Requesting 50 results for the combination to populate the actual top 10 URLs should be a sufficient number to always find a matching title. For the edge-case that the secondary request does not have any result for a URL from the primary request, we'll use "(unknown)" as a fallback title to display in the table.

The above approach ensures that:
1. The numbers in the table are fully accurate and match the Analytics frontend, and there are no duplicates because of different titles.
2. We still display titles above every result, which is a nicer user experience than what the Analytics frontend shows.

**Comment by felixarntz** [comment]
To whoever will write the IB on this one:
> Since this whole logic of running a matching separate request and combining responses to associate titles with URLs is needed in all those widgets, it should be implemented in a reusable way.

Potential ideas:
1. Write a JS function that processes the second response and returns a map of `url => title` pairs. Then the widget has to look up the title there.
2. Invent our own Site Kit-specific Analytics metric, e.g. `gask:pageTitle` and add code on our Analytics module server-side to internally make that second request to get URLs and titles, then amend the primary response as if `gask:pageTitle` was a real Analytics metric that had been requested.

The first would be a pragmatic solution which is itself rather uncomplicated but also would probably end up in a bunch of duplicate code across all these widgets. The second solution would move the complexity to one spot on the server-side (and that implementation may take longer initially), but the end result would be seamless, as if the Analytics API just supported that field.

@aaemnnosttv Thoughts?

**Comment by aaemnnosttv** [comment]
@felixarntz I think making the second request internally sounds better to me. For one, we could actually do the two requests in the same batch. We can do this as long as the two requests have the same dateRanges, viewId, segments, samplingLevel, and cohortGroup. However, this would have the risk of a result potentially not including the title (although unlikely).

One variation would be to make the `ga:pagePath`-only request first and then make the second request with an added dimension filter for the `ga:pagePath` to only return results that are `IN_LIST` of those that were returned from the first request. This should (I think) guarantee that we always have a title for each row. This would require that the requests are done in series rather than parallel but given this would happen on the server, it would still be far more performant than doing on the client.

**Comment by felixarntz** [comment]
@aaemnnosttv Sounds good. I've added to the ACs this high-level approach, specifically that it should be handled "automatically" on the server-side using a custom Site Kit-specific Analytics metric.

**Comment by eugene-manuilov** [comment]
@felixarntz @aaemnnosttv I have added IB which is different from what is asked in AC. I think that introducing a new fake dimension and tailoring the getReport functionality on the backend to support it, will significantly increase the complexity of the already non-simple implementation of the `GET:report` datapoint. Let me know what you think.

**Comment by felixarntz** [comment]
@eugene-manuilov I'm not sure about the approach proposed in the IB. This is what we actually started from when thinking it through (see e.g. https://github.com/google/site-kit-wp/issues/3070#issuecomment-858778436), but per conversation with @aaemnnosttv we decided it makes more sense to solve it internally.

While it _does_ increase complexity of the server-side implementation, it solves the problem at its root. The approach in the IB fixes in a way that is component-specific, which could be fine if it was a very specific problem, but getting the title back for URLs is a general Site Kit feature that we may need in other widgets too (e.g. top AdSense pages etc. as well). That's why we should focus on a DRY solution - what you're proposing in the IB IMO just moves the complexity to the client. It is probably quicker to get that implementation done, but then we'll need to do the same change in every similar widget later. Fixing it on the server abstracts away the complexity, so that the React component code remains easily understandable.

**Comment by eugene-manuilov** [comment]
@felixarntz ok, how about creating a special selector in the analytics datastore that will return report data for pages? Let's say `getReportForPages( { metrics, limit } )`? It will do the same trick as described in IB but in the resolver. In this case, we won't repeat ourselves and won't pollute the handler for the `GET:report` datapoint on the server side.

**Comment by felixarntz** [comment]
@eugene-manuilov That could work. What do you envision the return format etc. to be? Would people need to manually specify the fake metric in `metrics`, or would it automatically be included? My concern with the latter approach would be that you may be unaware of the exact metrics the response returns, because:
* The new "fake" metric should appear in the response as if it was a metric.
* If we handle it on the server-side, it works just like it normally would: If you pass the fake metric, the server-side does its magic and returns it in the format you would expect.
* If we handle it in a specific selector that adds it to the response even if you didn't specifically request the metric that would be slightly confusing.

Rather than adding a new selector, how about we add logic in the `getReport` selector itself that takes care of it _if_ the fake metric is requested?

**Comment by eugene-manuilov** [comment]
@felixarntz I didn't mean to use fake metrics or dimensions. My suggestion was to add a higher-order selector that works like the regular `getReport` selector, but also makes an additional request to retrieve page titles from the Analytics Report API.

See what I mean here:

```javascript
// assets/js/modules/analytics/datastore/report.js

const baseActions = {
    receiveReportForPages( options, report ) {
        return {
            type: RECEIVE_REPORT_FOR_PAGES,
            payload: {
                report,
                options,
            }
        };
    }
};

function baseReducer( state, { type, payload } ) {
    switch( type ) {
        case RECEIVE_REPORT_FOR_PAGES:
            return {
                ...state, 
                reportsForPages: {
                    ...state.reportsForPages,
                    [ stringifyObject( options ) ]: report,
                },
             }; 
    }        
}

const baseResolvers = {
    *getReportForPages( options ) {
        // 1. get registy
        // 2. call getReport( { ...options, dimensions: [ 'ga:pagePath' ] } )
        // 3. wait till we get results from the first getReport call
        // 4. call getReport( {
        //         ...options,
        //         dimensions: [ 'ga:pagePath', 'ga:pageTitle' ],
        //         dimensionFilter: { 'ga:pagePath': [ ...pages from the first call... ] },
        //         limit: ( options.limit || 10 ) * 5,
        //     } )
        // 5. wait till we get results from the second getReport call 
        // 6. combine results into one report
        // 7. call receiveReportForPages( options, report ) action
    }
};

const baseSelectors = {
    getReportForPages( state, options ) {
        const { reportsForPages } = state;
        return reportsForPages[ stringifyObject( options ) ];
    }
}
```

```diff
 function DashboardPopularPagesWidget( { Widget, WidgetReportZero, WidgetReportError } ) {
     ...
     const { ... } = useSelect( ( select ) => {
         const args = {
             startDate,
             endDate,
-            dimensions: [
-                'ga:pageTitle',
-                'ga:pagePath',
-             ],
             metrics: [
                 {
                     expression: 'ga:pageviews',
                     alias: 'Pageviews',
                 },
             ],
             orderby: [
                 {
                     fieldName: 'ga:pageviews',
                     sortOrder: 'DESCENDING',
                 },
             ],
             limit: 10,
         };

         return {
             ...
-            data: store.getReport( args ),
-            error: store.getErrorForSelector( 'getReport', [ args ] ),
-            loading: ! store.hasFinishedResolution( 'getReport', [ args ] ),
+            data: store.getReportForPages( args ),
+            error: store.getErrorForSelector( 'getReportForPages', [ args ] ),
+            loading: ! store.hasFinishedResolution( 'getReportForPages', [ args ] ),
         };
     };

     return (
         ...
     );
 }
```

**Comment by felixarntz** [comment]
@eugene-manuilov I understand, my point is how do you envision the resulting response (the combined response) to look like? If you combine them, my previous point is still relevant, which is that the selector would basically return a response with a shape different than what one might expect given the dimensions and metrics passed.

**Comment by aaemnnosttv** [comment]
Chiming in here – one thing I like about solving it on the server-side is that there's less risk of the response getting out of sync with that of a real report since we'd be working with models rather than just matching the expected structure in JS.

**Comment by eugene-manuilov** [comment]
@felixarntz I have updated IB based on our discussion yesterday. Please, let me know if it looks good now.

**Comment by felixarntz** [comment]
@eugene-manuilov IB mostly lgtm, just a few things:

About the `getPageTitles` selector, we may also need to pass a few additional arguments to the `getReport` call, for example these calls always require start and end dates, and we should ensure those are the same as used for the other report API request with the main data.

> The `dimensionFilter` map should contain `ga:pageTitle => {array of incoming page paths}`

Probably a typo? I think this should have `ga:pagePath` instead of `ga:pageTitle`.

**Comment by eugene-manuilov** [comment]
> About the getPageTitles selector, we may also need to pass a few additional arguments to the getReport call, for example these calls always require start and end dates, and we should ensure those are the same as used for the other report API request with the main data.

Yes! Added it to IB.

> Probably a typo? I think this should have ga:pagePath instead of ga:pageTitle.

Good catch. Updated!

**Comment by tofumatt** [comment]
As we discussed, we want to solve this on the client-side, which is what the IB proposes. But the ACs still read:

> Since this whole logic of running a matching separate request and combining responses to associate titles with URLs is needed in all those widgets, it should be implemented in a reusable way on the server-side (i.e. the client-side should not have to trigger two separate requests). This can work as follows:

We should probably update that to avoid confusion.

This IB looks good, but there's a lack of clarity to me around how we choose which page title to display. We discussed this on a call, but I think the most recent available page titles (either from the Analytics API or from WordPress) would make sense. I think here you're proposing we select the most recent page title returned from Analytics, but I'm not sure.

Or do you think we'll just select the first one returned?

**Comment by eugene-manuilov** [comment]
> Or do you think we'll just select the first one returned?

Yes, just the first one returned is enough. Added it to IB.

**Comment by tofumatt** [comment]
Sounds good. IB ✅ 

(I amended the ACs to not require the implementation to be server-side, as we discussed on a call.)

**Comment by ivankruchkoff** [comment]
@eugene-manuilov resolved all CR issues, however there is an e2e failure, not sure what is triggering it though. Any ideas?

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

Currently in the process of testing this.  I have setup a [test site](https://stuck-civet.jurassic.ninja/)  with pages and posts.

[Post 1](https://stuck-civet.jurassic.ninja/2021/09/20/page-1/) / [Post 2](https://stuck-civet.jurassic.ninja/2021/09/20/post-2/) /  [Page 1](https://stuck-civet.jurassic.ninja/page-1/) 

Once Analytics is updated in Site Kit, and the traffic recorded, I will go in and change the titles to QA this ticket.

**Comment by aaemnnosttv** [comment]
@wpdarren let's see if we can find a site that already has such posts as waiting for it to show up in Analytics and then changing the titles plus additional waiting will likely take too long.

@jamesozzie is there a site you're aware of that we could use to test this with? If so, we'd only need access to the Analytics view to test it.

**Comment by jamesozzie** [comment]
@aaemnnosttv I've changed the title of the contact page and the slug for the typography page from the below site - which I've grant access to now. I've only just changed both, so it will likely be tomorrow before they appear in the top content table (I've ensured a few visits so they do appear)

![image](https://user-images.githubusercontent.com/41326532/134178976-be0973a8-e3d0-4a5f-b762-2a0bffa35250.png)

**Comment by wpdarren** [comment]
@aaemnnosttv good point, I assumed that it was 24 hours for the Analytics to be updated but thats not looking to be the case. Thanks @jamesozzie please could you grant me access to that site and I can check it over tomorrow.

**Comment by jamesozzie** [comment]
@wpdarren Not a problem, I provided access so you should have an email invite.

**Comment by jamesozzie** [comment]
@wpdarren @aaemnnosttv That site is now displaying those URLs in the most popular table as per the screengrab below. You should have access to that property view. Let me know if you need a wp-admin login

![image](https://user-images.githubusercontent.com/41326532/134313284-4d7d7551-ce11-4b07-b10c-42f22f6db04b.png)

**Comment by wpdarren** [comment]
### QA Update ⚠️ 

@aaemnnosttv @jamesozzie when I am on the `develop` branch, for the page with the title change, I only see one entry, but the title is not the latest title, i.e. /contact the title is `Contact – new title test` but I am seeing just `Contact` This is on all of the URLs in the QAB.  Could you clarify what the title should appear as? 

![image](https://user-images.githubusercontent.com/73545194/134356416-16216255-68cb-4c7a-b28e-0bd8e928ab2f.png)

**Comment by aaemnnosttv** [comment]
@wpdarren from what I can see we're getting both new and old titles back and our logic is to pick the first one, it's just that the first one isn't the one we would expect it to be (the most recent). 

![image](https://user-images.githubusercontent.com/1621608/134472083-de64737f-7780-4df5-bb98-87160f4e9cd9.png)

This may be a matter of timing still because the Analytics API is not yet returning data for 9/23 which only has data with the new title whereas 9/21 (no data for this page from 9/22) has rows with both titles.

![image](https://user-images.githubusercontent.com/1621608/134472215-7f0e88f2-5c16-4298-a54b-eab0dd14bcca.png)

Still, our logic currently requires/assumes the newer data would be returned first but that doesn't seem to be the case. As far as I can tell, the page titles report data is coming back sorted ascending alphabetically by page path.
![image](https://user-images.githubusercontent.com/1621608/134472481-67479dec-9504-41b1-b53d-cf1e9636266d.png)

If this is the case, the solution may be to add the date dimension and an orderby statement to sort by the date in reverse descending order, which I've tested, but because no data is returned for today yet from the API we only have the one day which has data recorded with both titles so it isn't clear whether this would work or not (although I think it would). It would result in a much larger report returned for page titles though (i.e. all the requested page paths, for every day in the range) which might be a problem, particularly for larger date ranges.

**Comment by wpdarren** [comment]
Thanks @aaemnnosttv 👍 

Just to add from a conversation with Evan on Slack, I am going to check back on the site tomorrow to see if the API has updated the title and will report back. Will update when there's more information.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

This ticket passes the QAB `Ensure each url is entered only once, unlike in the screenshot in bug description.`

The URL only appears once when the title is changed.  

@aaemnnosttv @ivankruchkoff  I will create a new ticket for the issue because the new title does not appear, and we have waited a few days for the API to change this. The visits have registered.

**Comment by aaemnnosttv** [comment]
@wpdarren @felixarntz looking at this again today with the fresher data, it's still not choosing the most recent version of the title. Perhaps that's not the most important thing in this issue and probably shouldn't block it at this point in the release, but if that is something we want to ensure going forward we can fix it by including the `ga:date` dimension as mentioned above. I'll open a new issue for this as a follow-up.
